### ISSUE TITLE
[NFR] Add an option to cancel other validators only for the same field when the validator fails

### ISSUE THREAD

**Issue by superbull** [phalcon/cphalcon#1540]
For example:

``` PHP
<?php

use Phalcon\Validation\Validator\PresenceOf,
    Phalcon\Validation\Validator\Regex;

$validation = new Phalcon\Validation();

$validation
    ->add('telephone', new PresenceOf(array(
        'message' => 'The telephone is required',
        'cancelOnFail' => true
    )))
    ->add('telephone', new Regex(array(
        'message' => 'The telephone is required',
        'pattern' => '/\+44 [0-9]+/'
    )))
    ->add('telephone', new StringLength(array(
        'minimumMessage' => 'The telephone is too short',
        'min' => 2
    )))
    ->add('email', new PresenceOf(array(
        'message' => 'The email is required'
    )));
```

If the validator PresenceOf for 'telephone' fails, the validator execution status:

<table>
<tr>
  <td>telephone:PresenceOf</td>
  <td>executed</td>
</tr>
<tr>
  <td>telephone:Regex</td>
  <td>not executed</td>
</tr>
<tr>
  <td>telephone:StringLength</td>
  <td>not executed</td>
</tr>
<tr>
  <td>email:PresenceOf</td>
  <td>not executed</td>
</tr>
</table>

Sometimes we may want to restrict the **cancelOnFail** behavior to only one field.  Possible syntax:

``` PHP
<?php

use Phalcon\Validation\Validator\PresenceOf,
    Phalcon\Validation\Validator\Regex;

$validation = new Phalcon\Validation();

$validation
    ->add('telephone', new PresenceOf(array(
        'message' => 'The telephone is required',
        'cancelFieldValidationOnFail' => true
    )))
    ->add('telephone', new Regex(array(
        'message' => 'The telephone is required',
        'pattern' => '/\+44 [0-9]+/'
    )))
    ->add('telephone', new StringLength(array(
        'minimumMessage' => 'The telephone is too short',
        'min' => 2
    )))
    ->add('email', new PresenceOf(array(
        'message' => 'The email is required'
    )));
```

We use the new option **cancelFieldValidationOnFail** (may be you have better name for this). This time, if PresenceOf for 'telephone' fails, the validator execution status is:

<table>
<tr>
  <td>telephone:PresenceOf</td>
  <td>executed</td>
</tr>
<tr>
  <td>telephone:Regex</td>
  <td>not executed</td>
</tr>
<tr>
  <td>telephone:StringLength</td>
  <td>not executed</td>
</tr>
<tr>
  <td>email:PresenceOf</td>
  <td>executed</td>
</tr>
</table>

I think this option is useful.
##

**Comment by sergeyklay** [comment]
+1

**Comment by hi2u** [comment]
I'm actually already seeing the behavior that's being requested in this issue.  So I though that was how it works already.  i.e. cancelOnFail only cancels remaining validators on its own field, not the other fields.

Here's my login form code...

```php
class LoginForm extends \Phalcon\Forms\Form
{
    public function initialize()
    {
        $e = new Text('login');
        $e->setLabel('Email or username');
        $e->addValidators([
            new PresenceOf(array(
                'message' => 'Please enter your username or email address',
                'cancelOnFail' => true
            )),
            new LoginAllowValidator([
                'Form' => $this
            ]),
        ]);
        $this->add($e);
        
        // Password
        $password = new Password('currentPassword');
        $password->setLabel('Password');
        $password->addValidators(array(
            new PresenceOf(array(
                'message' => 'Please enter your current password',
                'cancelOnFail' => true
            )),
            new PasswordCheckValidator([
                'Form' => $this
            ]),
        ));
        $this->add($password);
    }
}
```

Even when the any of login's cancelOnFail validators fail, the password validators are still checked in the above code.  Could this be because I'm running ```$(fieldname)->addValidators()``` separately on each field?  Are they considered different "validation chains" or something like that?

The documentation at https://docs.phalconphp.com/en/latest/reference/validation.html#cancelling-validations is a little bit ambiguous, as the example only shows a single "telephone" field.

> therefore if that validator fails the remaining validators in the chain are not executed.

I'm not sure what "chain" means exactly, but I'm new to using Phalcon's validation.  Is my code example above 1 or 2 "chains"?

**Comment by Jurigag** [comment]
Im guessing we could change in 4.0.0 `cancelOnFail` to cancel whole validation and add proposed `cancelFieldValidationOnFail`

**Comment by niden** [comment]
Closing in favor of #13855. Will revisit if the community votes for it, or in later versions.

**Comment by Jeckerson** [comment]
Implemented in v4.0
